optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsGroupingBiConstraintStream.java

```diff
@@ -0,0 +1,104 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import org.drools.model.Global;
+import org.drools.model.PatternDSL;
+import org.drools.model.Rule;
+import org.drools.model.RuleItemBuilder;
+import org.optaplanner.core.api.score.holder.AbstractScoreHolder;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraint;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAbstractConstraintStream;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsLogicalTuple;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsAbstractUniConstraintStream;
+
+import static org.drools.model.DSL.declarationOf;
+import static org.drools.model.PatternDSL.pattern;
+
+public class DroolsGroupingBiConstraintStream<Solution_, A, NewA, ResultContainer_, NewB>
+        extends DroolsAbstractBiConstraintStream<Solution_, NewA, NewB> {
+
+    private final DroolsAbstractUniConstraintStream<Solution_, A> parent;
+    private final Function<A, NewA> groupKeyMapping;
+    private final UniConstraintCollector<A, ResultContainer_, NewB> collector;
+    private final AtomicInteger ruleId = new AtomicInteger(-1);
+
+    public DroolsGroupingBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractUniConstraintStream<Solution_, A> parent, Function<A, NewA> groupKeyMapping,
+            UniConstraintCollector<A, ResultContainer_, NewB> collector) {
+        super(constraintFactory);
+        this.parent = parent;
+        this.groupKeyMapping = groupKeyMapping;
+        this.collector = collector;
+    }
+
+    @Override
+    public Optional<Rule> buildRule(DroolsConstraint<Solution_> constraint,
+            Global<? extends AbstractScoreHolder<?>> scoreHolderGlobal) {
+        Object createdRuleId = createRuleIdIfAbsent(constraint.getConstraintFactory());
+        String ruleName = "Helper rule #" + createdRuleId + " (GroupBy remapping+collecting)";
+        Rule rule = PatternDSL.rule(constraint.getConstraintPackage(), ruleName)
+                .build(parent.createCondition().completeWithLogicalInsert(createdRuleId, groupKeyMapping, collector)
+                        .toArray(new RuleItemBuilder<?>[0]));
+        return Optional.of(rule);
+    }
+
+    @Override
+    public DroolsBiCondition<NewA, NewB> createCondition() {
+        return new DroolsBiCondition<>(declarationOf(DroolsLogicalTuple.class), var -> {
+            Object createdRuleId = createRuleIdIfAbsent(getConstraintFactory());
+            return pattern(var).expr(logicalTuple ->
+                    Objects.equals(logicalTuple.getRuleId(), createdRuleId));
+        });
+    }
+
+    /**
+     * Exists in order to be able to bind the condition in the new rule to the {@link DroolsLogicalTuple}s generated by
```

Comment:
nitpick style: Instead of "Exists in order to ..." I 'd use something like "Guarantees that ..."